File: starterAIs/starter.rb
class Cell
  def initialize(index, richness, neighbors)
    @index = index
    @richness = richness
    @neighbors = neighbors
  end
end

class Tree
  def initialize(cell_index, size, is_mine, is_dormant)
    @cell_index = cell_index
    @size = size
    @is_mine = is_mine
    @is_dormant = is_dormant
  end
end

WAIT = 'WAIT'
SEED = 'SEED'
GROW = 'GROW'
COMPLETE = 'COMPLETE'

class Action
  def self.parse(line)
    parts = line.split ' '
    if (parts[0] == WAIT)
      return Action.new(WAIT)
    end
    if (parts[1] == SEED)
      return Action.new(SEED, parts[2].to_i, parts[1].to_i)
    end
    return Action.new(parts[0], parts[1].to_i)
  end

  def initialize(type, target_cell_idx = 0, source_cell_idx = 0)
    @type = type
    @target_cell_idx = target_cell_idx
    @source_cell_idx = source_cell_idx
  end

  def to_s
    return WAIT if @type == WAIT
    return "#{SEED} #{@source_cell_idx} #{@target_cell_idx}" if @type == SEED
    "#{@type} #{@target_cell_idx}"
  end
end

class Game
  attr_accessor :day,
                :nutrients,
                :cells,
                :possible_actions,
                :trees,
                :my_sun,
                :my_score,
                :opponent_sun,
                :opponent_score,
                :opponent_is_waiting

  def initialize
    @day = 0
    @nutrients = 0
    @cells = []
    @possible_actions = []
    @trees = []
    @my_sun = 0
    @my_score = 0
    @opponent_sun = 0
    @opponent_score = 0
    @opponent_is_waiting = 0
  end

  def get_next_action
    # TODO: write your algorithm here
    return @possible_actions[0]
  end
end

game = Game.new

STDOUT.sync = true # DO NOT REMOVE
##  the standard input according to the problem statement.

number_of_cells = gets.to_i # 37
number_of_cells.times do
  # index: 0 is the center cell, the next cells spiral outwards
  # richness: 0 if the cell is unusable, 1-3 for usable cells
  # neigh_0: the index of the neighbouring cell for each direction
  index, richness, neigh_0, neigh_1, neigh_2, neigh_3, neigh_4, neigh_5 = gets.split(' ').collect { |x| x.to_i }
  game.cells << Cell.new(index, richness, [neigh_0, neigh_1, neigh_2, neigh_3, neigh_4, neigh_5])
end

# game loop
loop do
  game.day = gets.to_i # the game lasts 24 days: 0-23
  game.nutrients = gets.to_i # the base score you gain from the next COMPLETE action

  # my_sun: your sun points
  # my_score: your current score
  game.my_sun, game.my_score = gets.split(' ').collect { |x| x.to_i }

  # opponent_sun: opponent's sun points
  # opponent_score: opponent's score
  # opponent_is_waiting: whether your opponent is asleep until the next day
  inputs = gets.split(' ')
  game.opponent_sun = inputs[0].to_i
  game.opponent_score = inputs[1].to_i
  game.opponent_is_waiting = inputs[2] != '0'

  game.trees = []
  number_of_trees = gets.to_i # the current amount of trees
  number_of_trees.times do
    # cell_index: location of this tree
    # size: size of this tree: 0-3
    # is_mine: 1 if this is your tree
    # is_dormant: 1 if this tree is dormant
    cell_index, size, is_mine, is_dormant = gets.split ' '
    cell_index = cell_index.to_i
    size = size.to_i
    is_mine = is_mine.to_i == 1
    is_dormant = is_dormant.to_i == 1
    game.trees << Tree.new(cell_index, size, is_mine, is_dormant)
  end

  game.possible_actions = []
  number_of_possible_actions = gets.to_i
  number_of_possible_actions.times do
    possible_action = gets.chomp
    game.possible_actions << Action.parse(possible_action)
  end

  action = game.get_next_action

  puts action.to_s
end

File: starterAIs/starter.dart
/// This a starter AIs to help new player written in Dart
/// Use Dart version 2.12.4 which supports sound null safety.
import 'dart:io';

void debug(Object obj) {
  stderr.writeln(obj);
}

class Cell {
  final int index;
  final int richness;
  final List<int> neighbors;

  Cell(this.index, this.richness, this.neighbors);
}

class Tree {
  final int index;
  final int size;
  final bool isMine;
  final bool isDormant;

  Tree(this.index, this.size, this.isMine, this.isDormant);
}

class Action {
  static const String wait = 'WAIT';
  static const String seed = 'SEED';
  static const String grow = 'GROW';
  static const String complete = 'COMPLETE';

  final String type;
  final int? sourceCellIdx;
  final int? targetCellIdx;

  Action(
    this.type, [
    this.targetCellIdx,
    this.sourceCellIdx,
  ]);

  static parse(String action) {
    debug(action);
    List<String> parts = action.split(' ');

    switch (parts[0]) {
      case wait:
        return Action(wait);
      case seed:
        return Action(seed, int.parse(parts[2]), int.parse(parts[1]));
      default:
        return Action(parts[0], int.parse(parts[1]));
    }
  }

  @override
  String toString() {
    if (type == wait) {
      return wait;
    } else if (type == seed) {
      return '$seed $sourceCellIdx $targetCellIdx';
    } else {
      return '$type $targetCellIdx';
    }
  }
}

class Game {
  List<Cell> board = [];
  List<Action> possibleActions = [];
  List<Tree> trees = [];
  late int day;
  late int nutrients;
  late int mySun, opponentSun;
  late int myScore, opponentScore;
  late bool opponentIsWaiting;

  Game();

  Action getNextAction() {
    // TODO: write your algorithm here
    return this.possibleActions[0];
  }
}

String readLineSync() {
  String? s = stdin.readLineSync();
  return s == null ? '' : s;
}

/**
 * Auto-generated code below aims at helping you parse
 * the standard input according to the problem statement.
 **/
void main() {
  Game game = Game();

  List inputs;
  int numberOfCells = int.parse(readLineSync()); // 37
  for (int i = 0; i < numberOfCells; i++) {
    inputs = readLineSync().split(' ');
    int index = int.parse(
        inputs[0]); // 0 is the center cell, the next cells spiral outwards
    int richness =
        int.parse(inputs[1]); // 0 if the cell is unusable, 1-3 for usable cells
    int neigh0 = int.parse(
        inputs[2]); // the index of the neighbouring cell for each direction
    int neigh1 = int.parse(inputs[3]);
    int neigh2 = int.parse(inputs[4]);
    int neigh3 = int.parse(inputs[5]);
    int neigh4 = int.parse(inputs[6]);
    int neigh5 = int.parse(inputs[7]);

    List<int> neighs = [neigh0, neigh1, neigh2, neigh3, neigh4, neigh5];
    Cell cell = Cell(index, richness, neighs);

    game.board.add(cell);
  }

  // game loop
  while (true) {
    game.day = int.parse(readLineSync()); // the game lasts 24 days: 0-23
    game.nutrients = int.parse(
        readLineSync()); // the base score you gain from the next COMPLETE action
    inputs = readLineSync().split(' ');
    game.mySun = int.parse(inputs[0]); // your sun points
    game.myScore = int.parse(inputs[1]); // your current score
    inputs = readLineSync().split(' ');
    game.opponentScore = int.parse(inputs[0]); // opponent's sun points
    game.opponentScore = int.parse(inputs[1]); // opponent's score
    game.opponentIsWaiting = int.parse(inputs[2]) !=
        0; // whether your opponent is asleep until the next day

    game.trees.clear();

    int numberOfTrees =
        int.parse(readLineSync()); // the current amount of trees
    for (int i = 0; i < numberOfTrees; i++) {
      inputs = readLineSync().split(' ');
      int cellIndex = int.parse(inputs[0]); // location of this tree
      int size = int.parse(inputs[1]); // size of this tree: 0-3
      bool isMine = int.parse(inputs[2]) != 0; // 1 if this is your tree
      bool isDormant = int.parse(inputs[3]) != 0; // 1 if this tree is dormant
      Tree tree = Tree(cellIndex, size, isMine, isDormant);
      game.trees.add(tree);
    }

    game.possibleActions.clear();

    int numberOfPossibleActions =
        int.parse(readLineSync()); // all legal actions
    for (int i = 0; i < numberOfPossibleActions; i++) {
      String possibleAction =
          readLineSync(); // try printing something from here to start with
      game.possibleActions.add(Action.parse(possibleAction));
    }

    // Write an action using print()
    // To debug: stderr.writeln('Debug messages...');

    // GROW cellIdx | SEED sourceIdx targetIdx | COMPLETE cellIdx | WAIT <message>
    Action action = game.getNextAction();
    print(action);
  }
}
